File: 2022/2/day2.py
from sys import stdin

moves = {
    'A': 'ROCK',
    'B': 'PAPER',
    'C': 'SCISSORS',
    'X': 'ROCK',
    'Y': 'PAPER',
    'Z': 'SCISSORS',
}

scores = {
    'ROCK': 1,
    'PAPER': 2,
    'SCISSORS': 3,
    'WIN': 6,
    'TIE': 3,
    'LOSS': 0,
}

def score(a, b):
    if moves[a] == moves[b]:
        return scores['TIE'] + scores[moves[b]]

    if moves[a] == 'ROCK':
        if moves[b] == 'PAPER':
            return scores['WIN'] + scores[moves[b]]
        if moves[b] == 'SCISSORS':
            return scores['LOSS'] + scores[moves[b]]
    if moves[a] == 'PAPER':
        if moves[b] == 'ROCK':
            return scores['LOSS'] + scores[moves[b]]
        if moves[b] == 'SCISSORS':
            return scores['WIN'] + scores[moves[b]]
    if moves[a] == 'SCISSORS':
        if moves[b] == 'ROCK':
            return scores['WIN'] + scores[moves[b]]
        if moves[b] == 'PAPER':
            return scores['LOSS'] + scores[moves[b]]

def calculate_move(a, b):
    results = {
        'X': 'LOSS',
        'Y': 'TIE',
        'Z': 'WIN',
    }

    if moves[a] == 'ROCK':
        if results[b] == 'LOSS':
            return 'Z'
        if results[b] == 'TIE':
            return 'X'
        if results[b] == 'WIN':
            return 'Y'
    if moves[a] == 'PAPER':
        if results[b] == 'LOSS':
            return 'X'
        if results[b] == 'TIE':
            return 'Y'
        if results[b] == 'WIN':
            return 'Z'
    if moves[a] == 'SCISSORS':
        if results[b] == 'LOSS':
            return 'Y'
        if results[b] == 'TIE':
            return 'Z'
        if results[b] == 'WIN':
            return 'X'

def main():
    rounds = [line.strip().split(' ') for line in stdin]

    print(sum([score(r[0], r[1]) for r in rounds]))
    print(sum([score(r[0], calculate_move(r[0], r[1])) for r in rounds]))

if __name__ == '__main__':
    main()

File: 2022/2/day2.exs
defmodule Day2 do
  def score("A", "X"), do: 3 + 1
  def score("A", "Y"), do: 6 + 2
  def score("A", "Z"), do: 0 + 3
  def score("B", "X"), do: 0 + 1
  def score("B", "Y"), do: 3 + 2
  def score("B", "Z"), do: 6 + 3
  def score("C", "X"), do: 6 + 1
  def score("C", "Y"), do: 0 + 2
  def score("C", "Z"), do: 3 + 3

  def calculate_move("A", "X"), do: "Z"
  def calculate_move("A", "Y"), do: "X"
  def calculate_move("A", "Z"), do: "Y"
  def calculate_move("B", "X"), do: "X"
  def calculate_move("B", "Y"), do: "Y"
  def calculate_move("B", "Z"), do: "Z"
  def calculate_move("C", "X"), do: "Y"
  def calculate_move("C", "Y"), do: "Z"
  def calculate_move("C", "Z"), do: "X"
end

rounds =
  IO.read(:stdio, :eof)
  |> String.split("\n", trim: true)
  |> Enum.map(&String.split/1)

Enum.map(rounds, fn [a, b] -> Day2.score(a, b) end)
|> Enum.sum()
|> IO.puts()

Enum.map(rounds, fn [a, b] -> Day2.score(a, Day2.calculate_move(a, b)) end)
|> Enum.sum()
|> IO.puts()
